File: Cpp/factory_method.cpp
#include <iostream>
#include <memory>

class Document {
public:
    virtual ~Document() {}
    virtual std::string getType() = 0;
};

class PdfDocument : public Document {
public:
    std::string getType() override {
        return "PDF Document";
    }
};

class WordDocument : public Document {
public:
    std::string getType() override {
        return "Word Document";
    }
};

class DocumentFactory {
public:
    virtual ~DocumentFactory() {}
    virtual std::unique_ptr<Document> createDocument() = 0;
};

class PdfDocumentFactory : public DocumentFactory {
public:
    std::unique_ptr<Document> createDocument() override {
        return std::make_unique<PdfDocument>();
    }
};

class WordDocumentFactory : public DocumentFactory {
public:
    std::unique_ptr<Document> createDocument() override {
        return std::make_unique<WordDocument>();
    }
};

int main() {
    PdfDocumentFactory pdfFactory;
    auto doc1 = pdfFactory.createDocument();
    std::cout << doc1->getType() << std::endl;  // Prints: PDF Document

    WordDocumentFactory wordFactory;
    auto doc2 = wordFactory.createDocument();
    std::cout << doc2->getType() << std::endl;  // Prints: Word Document

    return 0;
}

File: Python/factory_method.py
from abc import ABC, abstractmethod


class Document(ABC):
    @abstractmethod
    def show_type(self):
        pass


class PdfDocument(Document):
    def show_type(self):
        return "PDF Document"


class WordDocument(Document):
    def show_type(self):
        return "Word Document"


class DocumentFactory(ABC):
    @abstractmethod
    def create_document(self) -> Document:
        pass


class PdfDocumentFactory(DocumentFactory):
    def create_document(self) -> Document:
        return PdfDocument()


class WordDocumentFactory(DocumentFactory):
    def create_document(self) -> Document:
        return WordDocument()


if __name__ == "__main__":
    # Użycie
    pdf_factory = PdfDocumentFactory()
    document = pdf_factory.create_document()
    print(document.show_type())  # Wyświetla: PDF Document

    word_factory = WordDocumentFactory()
    document = word_factory.create_document()
    print(document.show_type())  # Wyświetla: Word Document
